lib/vtls/schannel.c

```diff
@@ -907,20 +907,40 @@ schannel_connect_step1(struct Curl_cfilter *cf, struct Curl_easy *data)
 
   {
 #ifdef HAS_ALPN_SCHANNEL
-    bool wine;
+    static bool s_wine = FALSE;
+    static bool s_wine_has_alpn = FALSE;
+    bool os_has_alpn;
+
 #ifdef CURL_WINDOWS_UWP
     /* GetModuleHandle() not available for UWP.
        Assume no WINE because WINE has no UWP support. */
-    wine = FALSE;
+    s_wine = FALSE;
 #else
-    wine = !!GetProcAddress(GetModuleHandle(TEXT("ntdll")),
-                            "wine_get_version");
+    typedef const char *(APIENTRY *WINE_GET_VERSION_FN)(void);
+    static WINE_GET_VERSION_FN s_p_wine_get_version;
+    static bool s_wine_init = FALSE;
+    if(!s_wine_init) {
+      s_p_wine_get_version =
+        CURLX_FUNCTION_CAST(WINE_GET_VERSION_FN,
+                            (GetProcAddress(GetModuleHandle(TEXT("ntdll")),
+                                            "wine_get_version")));
+      s_wine = !!s_p_wine_get_version;
+      if(s_wine) {
+        const char *wine_version = s_p_wine_get_version(); /* e.g. "6.0.2" */
+        /* Assume ALPN support with WINE 6.0.0 or upper */
+        s_wine_has_alpn = wine_version && atoi(wine_version) >= 6;
+      }
+      s_wine_init = TRUE;
```

Comment:
Use of static variables in the library outside of its initialization is not good and we don't prefix them with `s_`. I would move this block into schannel_init and set just one static variable top at global scope `static bool os_has_alpn;` then back here you could just do a one liner `backend->use_alpn = connssl->alpn && os_has_alpn;`

Reply:
Good idea! Pushed an update to this effect.

Could we use the same method to avoid other repeat calls to `curlx_verify_windows_version()`? They look fairly costly and caching them may improve perf.

Reply:
We have global [Curl_isVistaOrGreater](https://github.com/curl/curl/blob/curl-8_12_1/lib/system_win32.c#L40) in win32 init so I suggest something simliar to that. For build numbers that's a little more specific so the scope may be narrower like for example schannel init could check if TLS 1.3 is supported by just Schannel (for example imagine multissl other backends may support it) so there's no reason to global that as Curl_. #16400 to get the ball rolling